### ISSUE TITLE
Rule proposal: `prefer-promise-all-settled`

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#1427]
Recently, I found I'm starting to use `Promise.allSettled()` instead of `Promise.all()`.
I only use `Promise.all()` when all promises are 100% sure safe to resolve, but I'm not sure where is the boundary.
I'd like to open up a discussion to see if we can do this safely.

## Fail

```js
const foo = await Promise.all(promises);
```

## Pass

```js
const foo = await Promise.allSettled(promises);
```

**Comment by fisker** [comment]
Maybe when we are checking the error thrown by `Promise.all()`?

**Comment by sindresorhus** [comment]
Often, you also want it to end early when a promise rejects though. I don't really see how we can statically detect this.

**Comment by sindresorhus** [comment]
> Maybe when we are checking the error thrown by Promise.all()? Means if we don't care about what the error is, Promise.all() should be preferred.

Not entirely sure what you mean. Maybe you could show some examples?

**Comment by fisker** [comment]
I was thinking about these, 

```js
// Good
try {
	await Promise.all(promises);
} catch {}
```

```js
// Bad
try {
	await Promise.all(promises);
} catch (error) {
	if (isAError(error)) {
	}
	if (isBError(error)) {
	}
	if (isCError(error)) {
	}
}
```

```js
const results = await Promise.allSettled(promises);
const errors = results.filter(({status}) => status === 'rejected');
```

but seems also not very obvious what to use.

**Comment by fisker** [comment]
What about this

```js
Promise.all(promises.map(promise => promise.catch(...)))
```

This case definitely should use `Promise.allSettled`.

**Comment by fisker** [comment]
> Often, you also want it to end early when a promise rejects though. I don't really see how we can statically detect this.

I'm aware of that, I opened this proposal to see if others get some cases for this rule.

**Comment by sindresorhus** [comment]
> Promise.all(promises.map(promise => promise.catch(...)))

That is a good case, I agree. Same with try/catch instead the map.

**Comment by sindresorhus** [comment]
This is accepted.

**Comment by mweinand** [comment]
It would be nice to have this rule as a suggestion - "Are you sure you understand the error handling tradeoff between .all and .allSettled?". I see a lot of people use .all when they don't realize one error will prevent the other promises from completing.

**Comment by sindresorhus** [comment]
Closing. `Promise.all()` and `Promise.allSettled()` express different semantics, and even the narrow `Promise.all(promises.map(promise => promise.catch(...)))` pattern is not safely replaceable in general because the catch callback may transform values, produce fallbacks, perform side effects, or return a result shape different from `allSettled()`.

A rule precise enough to avoid false positives would ned to analyze callback behavior and result usage, while the truly equivalent reflect pattern is too narrow to justify a Unicorn rule.
